examples/http_parallel_call/src/lib.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use proxy_wasm::hostcalls;
+use proxy_wasm::promise::Promise;
+use proxy_wasm::traits::*;
+use proxy_wasm::types::*;
+use std::collections::HashMap;
+use std::rc::Rc;
+use std::time::Duration;
+
+proxy_wasm::main! {{
+    proxy_wasm::set_log_level(LogLevel::Trace);
+    proxy_wasm::set_http_context(|_, _| -> Box<dyn HttpContext> { Box::new(HttpParallelCall::default()) });
+}}
+
+#[derive(Default)]
+struct HttpParallelCall {
+    m: HashMap<u32, Rc<Promise<(u32, usize, usize, usize)>>>,
+}
+
+impl HttpContext for HttpParallelCall {
+    fn on_http_request_headers(&mut self, _: usize, _: bool) -> Action {
+        // "Hello, "
+        let token1 = self
+            .dispatch_http_call(
+                "httpbin",
+                vec![
+                    (":method", "GET"),
+                    (":path", "/base64/SGVsbG8sIAo="),
+                    (":authority", "httpbin.org"),
+                ],
+                None,
+                vec![],
+                Duration::from_secs(1),
+            )
+            .unwrap();
+
+        // "World!"
+        let token2 = self
+            .dispatch_http_call(
+                "httpbin",
+                vec![
+                    (":method", "GET"),
+                    (":path", "/base64/V29ybGQhCg=="),
+                    (":authority", "httpbin.org"),
+                ],
+                None,
+                vec![],
+                Duration::from_secs(1),
+            )
+            .unwrap();
+
+        let promise1 = Promise::new();
+        let promise2 = Promise::new();
+        self.m.insert(token1, promise1.clone());
+        self.m.insert(token2, promise2.clone());
```

Comment:
This is leaking a lot of async machinery into the plugin's codebase, which means that it becomes public API that we need to maintain, and that users need to handle it themselves.

Instead, we should be handling this within the SDK.

Reply:
I have implemented a dispatcher to wrap `dispatch_http_call` in my project, but not submitted in proxy-wasm. Does any suggested style or file organized in proxy-wasm? Otherwise I will take it to `dispatcher.rs` to handle this within the SDK :)

```rust
#[derive(Default)]
pub struct HttpDispatcher {
    m: HashMap<u32, Rc<Promise<(u32, usize, usize, usize)>>>,
}

impl HttpDispatcher {
    pub fn dispatch(
        &mut self,
        upstream: &str,
        headers: Vec<(&str, &str)>,
        body: Option<&[u8]>,
        trailers: Vec<(&str, &str)>,
        timeout: Duration,
    ) -> Rc<Promise<(u32, usize, usize, usize)>> {
        let token =
            hostcalls::dispatch_http_call(upstream, headers, body, trailers, timeout).unwrap();
        let promise = Promise::new();
        self.m.insert(token, promise.clone());
        promise
    }

    pub fn callback(
        &mut self,
        _token_id: u32,
        _num_headers: usize,
        _body_size: usize,
        _num_trailers: usize,
    ) {
        let promise = self.m.remove(&_token_id);
        promise
            .unwrap()
            .fulfill((_token_id, _num_headers, _body_size, _num_trailers))
    }
}
```